raiden/channel/balance_proof.py

```diff
@@ -209,30 +209,38 @@ def get_known_unlocks(self):
             self.hashlock_unlockedlocks.itervalues(),
         )
 
+        tree = self.generate_merkle_tree();
+
         return [
             self.compute_proof_for_lock(
                 partialproof.secret,
                 partialproof.lock,
+                tree,
             )
             for partialproof in allpartialproof
         ]
 
-    def compute_proof_for_lock(self, secret, lock):
-        alllocks = chain(
-            self.hashlock_pendinglocks.values(),
-            self.hashlock_unclaimedlocks.values(),
-            self.hashlock_unlockedlocks.values()
-        )
+    def compute_proof_for_lock(self, secret, lock, tree=None):
```

Comment:
May I suggest to split this into two functions to avoid the `if` check?

The only "real" place in the code where this is used is inside `get_known_unlocks()` and due to your change the `tree` will always be given there.

All other cases where `compute_proof_for_lock` is used atm is in tests.

So I would suggest:

1. Create a function `compute_proof_for_lock_with_tree` and use it in `get_known_unlocks()`.
2. Leave `compute_proof_for_lock()` as it is functionality wise but code-wise it would be as simple as:

```python
def compute_proof_for_lock(self, secret, lock):
    tree = self.generate_merkle_tree()
    return self.compute_proof_for_lock_with_tree(secret, lock ,tree)
```